crates/core/src/rlp/decode.rs

```diff
@@ -49,174 +67,188 @@ impl RLPDecode for u8 {
 }
 
 impl RLPDecode for u16 {
-    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
-        let (bytes, _) = decode_bytes(rlp)?;
+    fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError> {
+        let (bytes, rest) = decode_bytes(rlp)?;
         let padded_bytes = static_left_pad(bytes)?;
-        Ok(u16::from_be_bytes(padded_bytes))
+        Ok((u16::from_be_bytes(padded_bytes), rest))
     }
 }
 
 impl RLPDecode for u32 {
-    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
-        let (bytes, _) = decode_bytes(rlp)?;
+    fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError> {
+        let (bytes, rest) = decode_bytes(rlp)?;
         let padded_bytes = static_left_pad(bytes)?;
-        Ok(u32::from_be_bytes(padded_bytes))
+        Ok((u32::from_be_bytes(padded_bytes), rest))
     }
 }
 
 impl RLPDecode for u64 {
-    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
-        let (bytes, _) = decode_bytes(rlp)?;
+    fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError> {
+        let (bytes, rest) = decode_bytes(rlp)?;
         let padded_bytes = static_left_pad(bytes)?;
-        Ok(u64::from_be_bytes(padded_bytes))
+        Ok((u64::from_be_bytes(padded_bytes), rest))
     }
 }
 
 impl RLPDecode for u128 {
-    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
-        let (bytes, _) = decode_bytes(rlp)?;
+    fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError> {
+        let (bytes, rest) = decode_bytes(rlp)?;
         let padded_bytes = static_left_pad(bytes)?;
-        Ok(u128::from_be_bytes(padded_bytes))
+        Ok((u128::from_be_bytes(padded_bytes), rest))
     }
 }
 
 impl<const N: usize> RLPDecode for [u8; N] {
-    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
-        let (decoded_bytes, _) = decode_bytes(rlp)?;
-        decoded_bytes
+    fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError> {
+        let (decoded_bytes, rest) = decode_bytes(rlp)?;
```

Comment:
I think there's some ambiguity here. `[u8; N]` could be interpreted as `[0x00, 0x01, ...]` or as `0x0001`. Here we're adopting the second one, but maybe the first one would be less confusing, since `Vec<u8>` is treated that way. We already have hash types (`H**`) for that anyways.

Reply:
comments added